### PR TITLE
Moving Import-PowerShellDatafile from script to compiled cmdlet

### LINKED ISSUE #2734: Rewrite Import-PowerShellDataFile on C#
As approved in #2718  Import-PowerShellDataFile should be rewritten on C#.

### PR DESCRIPTION
**Pull request by powercode**
Fixes #2734

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs [hunk 1]
```diff
@@ -0,0 +1,122 @@
+/********************************************************************++
+Copyright (c) Microsoft Corporation.  All rights reserved.
+--********************************************************************/
```
Comment (vors) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs]: @HemantMahawar should we have copyright preamble on the new files?

File: src/Microsoft.PowerShell.Commands.Utility/map.json [hunk 2]
```diff
@@ -47,6 +47,7 @@
     "monad/src/commands/utility/GetUnique.cs":  "commands/utility/GetUnique.cs",
     "monad/src/commands/utility/group-object.cs":  "commands/utility/group-object.cs",
     "monad/src/commands/utility/ImplicitRemotingCommands.cs":  "commands/utility/ImplicitRemotingCommands.cs",
+    "monad/src/commands/utility/ImportPowerShellDataFile.cs":  "commands/utility/ImportPowerShellDataFile.cs",
```
Comment (vors) [inline src/Microsoft.PowerShell.Commands.Utility/map.json]: That's probably not needed

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                    }
+                    else
+                    {
+                        WriteError(BuildInvalidDataFileErrorRecord(path, "CouldNotParseAsPowerShellDataFileNoHashtableRoot"));
+                    }
+                }
+            }
+        }
+
+        private List<string> ResolveFilePaths(string[] filePaths, bool isLiteralPath)
```
Comment (vors) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs]: This method is copied from https://github.com/PowerShell/PowerShell/blob/9a195f44c17773fdbb5a2e6a09de0c880072df23/src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs#L1723

Can we avoid such copy-paste?

**Review by vors — CHANGES_REQUESTED** [review]
Please, avoid copy-paste

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs (hunk 1 above)
Comment (HemantMahawar) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs]: Given it is open-source code, I don't think it is needed. @joeyaiello Can you weigh in?

File: src/Microsoft.PowerShell.Commands.Utility/map.json (hunk 2 above)
Comment (powercode) [inline src/Microsoft.PowerShell.Commands.Utility/map.json]: removed

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                    Token[] tokens;
+                    ParseError[] errors;
+                    var ast = Parser.ParseFile(resolved, out tokens, out errors);
+                    if (errors.Length > 0)
+                    {
+                        WriteInvalidDataFileError(resolved, String.Empty);
+                    }
+                    else
+                    {
+                        var data = ast.Find(a => a is HashtableAst, false);
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs]: I realize this is just a translation of the PowerShell implementation, but a psd1 can contain much more than a simple hashtable.  For example, the psd1 file contain 2 hashtables as an array, and this implementation would ignore the second one.

As implemented, the cmdlet is mostly useful for typical module manifests, but psd1 files are used for other things, e.g. localization, and sometimes use commands like `ConvertFrom-StringData`.

We should, as a minimum, not ignore any part of the ast.
We should also consider supporting more than a simple hashtable. `Ast.SafeGetValue` can certainly handle more (e.g. an array of hashtables).
We might also consider doing something more like what `Import-LocalizedData` does, though we'd need feedback from @LeeHolmes to understand the important differences and why this cmdlet was introduced.

**Review by lzybkr — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        /// </summary>
+        [Parameter(Mandatory = true,
+            Position = 0,
+            ParameterSetName = "ByLiteralPath")]
+        [ValidateNotNullOrEmpty()]
+        [SuppressMessage("Microsoft.Performance", "CA1819:PropertiesShouldNotReturnArrays")]
+        public string[] LiteralPath
+        {
+            get { return Path; }
+            set { _isLiteralPath = true; Path = value; }
```
Comment (powercode) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs]: A quick search gives that the pattern is in
```
AclCommands.cs
CertificateCommands.cs
CSVCommands.cs
ExportAliasCommand.cs
MatchString.cs
Out-File.cs
SaveHelpCommand.cs
SignatureCommands.cs
StartTranscriptCmdlet.cs
UpdateHelpCommand.cs
WriteFormatDataCommand.cs
XmlCommands.cs
```
But the cost of changing is very small.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs (hunk 4 above)
Comment (powercode) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs]: Should we start with just the move as descripted in the issue, and open a new one for the functional changes?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            var errorCategory = ErrorCategory.InvalidArgument;
+            var errorMessage = string.Format(UtilityResources.PathDoesNotExist, path);
+            var exception = new ArgumentException(errorMessage);
+            var errorRecord = new ErrorRecord(exception, errorId, errorCategory, path);
+            WriteError(errorRecord);
+        }
+
+        void WriteInvalidDataFileError(string resolvedPath, string extraError)
+        {
+            var errorId = $"CouldNotParseAsPowerShellDataFile{extraError}";
```
Comment (powercode) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs]: corrected.

Rebased version pushed.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    {
+        bool _isLiteralPath;
+
+        /// <summary>
+        /// Path specified, using globbing to resolve
+        /// </summary>
+        [Parameter(Mandatory = true,
+            Position = 0,
+            ParameterSetName = "ByPath",
+            ValueFromPipeline = true,
```
Comment (powercode) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs]: Corrected.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs (hunk 5 above)
Comment (powercode) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs]: either an extra field or
```
get { return _isLiteralPath ? Path : null; }
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs (hunk 4 above)
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs]: +1 don't block the PR and open new Issue.

**Comment by mirichmo** [comment]
@lzybkr and @vors Are your concerns resolved or is there additional work left here?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    {
+        bool _isLiteralPath;
+
+        /// <summary>
+        /// Path specified, using globbing to resolve
+        /// </summary>
+        [Parameter(Mandatory = true, Position = 0, ParameterSetName = "ByPath")]        
+        [ValidateNotNullOrEmpty]
+        [SuppressMessage("Microsoft.Performance", "CA1819:PropertiesShouldNotReturnArrays")]
+        public string[] Path { get; set; }
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImportPowerShellDataFile.cs]: We don't need a change as this code is just following a somewhat broken pattern.

If `LiteralPath` was set, `_isLiteralPath` will be set to true.  Then, if `Path` is set, `_isLiteralPath` is not cleared.  This isn't really a scenario I guess, but it is handled correctly in some places in our code base, but in many others it is not.

**Review by lzybkr — APPROVED** [review]

**Pull request merged**